api/paidAction/index.js

```diff
@@ -33,72 +36,162 @@ export const paidActions = {
+        [PAID_ACTION_PAYMENT_METHODS.FEE_CREDIT]: -2,
+        [PAID_ACTION_PAYMENT_METHODS.REWARD_SATS]: -1
+      }
+      supportedPaymentMethods = supportedPaymentMethods.sort((a, b) => {
+        return (priority[a] || 0) - (priority[b] || 0)
+      })
+    }
 
-      // payment methods that anonymous users can use
-      if (paymentMethod === PAID_ACTION_PAYMENT_METHODS.P2P) {
-        try {
-          return await performP2PAction(actionType, args, context)
-        } catch (e) {
-          if (!(e instanceof NonInvoiceablePeerError)) {
-            console.error(`${paymentMethod} action failed`, e)
-            throw e
+    for (const paymentMethod of supportedPaymentMethods) {
+      console.log('trying payment method', paymentMethod)
+      if (paymentMethod === PAID_ACTION_PAYMENT_METHODS.FEE_CREDIT) {
+        if (me && (me.mcredits ?? 0n) >= cost) {
+          try {
+            return await performInternalAction(cost, actionType, paidAction, paymentMethod, args, context)
+          } catch (e) {
+            console.error('payment method failed', paymentMethod, e)
           }
         }
-      } else if (paymentMethod === PAID_ACTION_PAYMENT_METHODS.PESSIMISTIC) {
-        return await performPessimisticAction(actionType, args, context)
-      }
-
-      // additionalpayment methods that logged in users can use
-      if (me) {
-        if (paymentMethod === PAID_ACTION_PAYMENT_METHODS.FEE_CREDIT ||
-          paymentMethod === PAID_ACTION_PAYMENT_METHODS.REWARD_SATS) {
+      } else if (paymentMethod === PAID_ACTION_PAYMENT_METHODS.REWARD_SATS) {
+        if (me && (me.msats ?? 0n) >= cost) {
           try {
-            return await performNoInvoiceAction(actionType, args, context, paymentMethod)
+            return await performInternalAction(cost, actionType, paidAction, paymentMethod, args, context)
           } catch (e) {
-            // if we fail with fee credits or reward sats, but not because of insufficient funds, bail
-            console.error(`${paymentMethod} action failed`, e)
-            if (!e.message.includes('\\"users\\" violates check constraint \\"mcredits_positive\\"') &&
-              !e.message.includes('\\"users\\" violates check constraint \\"msats_positive\\"')) {
-              throw e
-            }
+            console.error('payment method failed', paymentMethod, e)
           }
-        } else if (paymentMethod === PAID_ACTION_PAYMENT_METHODS.OPTIMISTIC) {
-          return await performOptimisticAction(actionType, args, context)
         }
+      } else if (paymentMethod === PAID_ACTION_PAYMENT_METHODS.P2P) {
+        try {
+          const optimistic =
+            paymentMethod === PAID_ACTION_PAYMENT_METHODS.OPTIMISTIC ||
+            // if action supports optimism and its a logged p2p action, we can try to be p2p and optimistic
+            (me && paymentMethod === PAID_ACTION_PAYMENT_METHODS.P2P &&
+              supportedPaymentMethods.includes(PAID_ACTION_PAYMENT_METHODS.OPTIMISTIC))
```

Comment:
This is to ensure p2p actions stay pessimistic unless the actions supports PAID_ACTION_PAYMENT_METHODS.OPTIMISTIC.